google-cloud-spanner/lib/google/cloud/spanner/interval.rb

```diff
@@ -0,0 +1,270 @@
+              raise ArgumentError, "The provided string does not follow ISO8601 standard."
+            end
+
+            if matches.captures.empty?
+              raise ArgumentError, "The provided string does not follow ISO8601 standard."
+            end
+
+            if matches[:years]
+              interval_months += matches[:years].to_i * 12
+            end
+
+            if matches[:months]
+              interval_months += matches[:months].to_i
+            end
+
+            if matches[:days]
+              interval_days = matches[:days].to_i
+            end
+
+            if matches[:hours]
+              interval_nanoseconds += matches[:hours].to_i * NANOSECONDS_IN_AN_HOUR
+            end
+
+            if matches[:minutes]
+              interval_nanoseconds += matches[:minutes].to_i * NANOSECONDS_IN_A_MINUTE
+            end
+
+            # Only seconds can be fractional. Both period and comma are valid inputs.
+            if matches[:seconds]
+              interval_nanoseconds += matches[:seconds].gsub(",", ".").to_f * NANOSECONDS_IN_A_SECOND
+            end
+
+            Interval.new interval_months, interval_days, interval_nanoseconds
+          end
+
+          # Returns an Interval instance with the given months.
+          #
+          # @param [Integer]
+          # @return [Interval]
+          def from_months months
+            Interval.new months, 0, 0
+          end
+
+          # Returns an Interval instance with the given days.
+          #
+          # @param [Integer]
+          # @return [Interval]
+          def from_days days
+            Interval.new 0, days, 0
+          end
+
+          # Returns an Interval instance with the given seconds.
+          #
+          # @param [Integer]
+          # @return [Interval]
+          def from_seconds seconds
+            nanoseconds = seconds_to_nanoseconds seconds
+            Interval.new 0, 0, nanoseconds
+          end
+
+          # Returns an Interval instance with the given milliseconds.
+          #
+          # @param [Integer]
+          # @return [Interval]
+          def from_milliseconds milliseconds
+            nanoseconds = milliseconds * NANOSECONDS_IN_A_MILLISECOND
+            Interval.new 0, 0, nanoseconds
+          end
+
+          # Returns an Interval instance with the given microseconds.
+          #
+          # @param [Integer]
+          # @return [Interval]
+          def from_microseconds microseconds
+            nanoseconds = microseconds * NANOSECONDS_IN_A_MICROSECOND
+            Interval.new 0, 0, nanoseconds
+          end
+
+          # Returns an Interval instance with the given nanoseconds.
+          #
+          # @param [Integer]
+          # @return [Interval]
+          def from_nanoseconds nanoseconds
+            Interval.new 0, 0, nanoseconds
+          end
+        end
+
+        def to_s
+          # Memoizing it as the logic can be a bit heavy.
+          @to_s ||= to_string
+        end
+
```

Comment:
Please also implement:

* `#eql?` / `#==` : these are generally synonymous, and usually one is implemented and the other is defined as an alias
* `#hash` : should be implemented whenever the above is implemented
* `#positive?`, `#negative?`, and `#zero?` just because I think they're useful for this class